### ISSUE TITLE
Myrtle Ave M/Z/J trains organized the wrong direction. 

### ISSUE THREAD

**Issue by lia-simeone** [jamespfennell/react-sandbox#28]
Huge fan of your web app. I've used it for several years in NYC. I recently moved to Bushwich Myrtle Ave M/Z/J stop. I kept missing the train to work and I couldn't figure out why.

This morning I realized your app had the direction of the M train reversed! The train listed under Manhattan is actually going away from Manhattan towards Brooklyn. The Manhattan-bound train is Forest Hills 71st. No wonder I kept missing it!

I haven't poked around in the source code, but I'd happy to make this change if I can find it. Thanks!

![266189B4-BE0F-4BA7-B864-C8000613A9F7](https://user-images.githubusercontent.com/11477059/186675309-b27a8d9d-d921-42d6-9797-5b581b77121a.png)

(Sorry hit create too soon)

**Comment by lia-simeone** [comment]
Sorry more thoughts. The reason I opened an issue here is because MTA real time webapp (which I think is fairly recent?) had the labels correct. Maybe that's not the same source as what you use?

![787E941B-AEF1-403B-8FE4-A37BCCADEA70](https://user-images.githubusercontent.com/11477059/186675801-ab2126fb-d8f7-4f85-8c8c-b5145ed328c3.png)

**Comment by jamespfennell** [comment]
Thank you so much for the bug report! And sorry to hear it caused you to miss some trains...

I think I know what's happening here. The raw data from the MTA for the M train in Williamsburg/Bushwick has a bug in which the directions are reported incorrectly. The old version of the backend code for the realtimerail app [has a manual fix for this bug](https://github.com/jamespfennell/transiter-ny/blob/5952f8c204358b271fc4abafba9aab1ca3787493/transiter_ny_mta/transiter_ny_mta/subwaytripsparser.py#L98-L107). Last week I migrated the app to use a new version of the backend code which was rewritten from scratch. It looks like I forgot to include that bug fix in the new code, and this caused this regression.

The new backend code parses the MTA's data using a [Go GTFS library I wrote](https://github.com/jamespfennell/gtfs), and this fix needs to be made in [this file](https://github.com/jamespfennell/gtfs/blob/main/extensions/nycttrips/nycttrips.go). I'll likely be able to do this today.

**Comment by lia-simeone** [comment]
Amazing thanks for the quick fix!

Can I ask you a question while I have you? It's not a bug so I didn't want to open an issue. Do you know this app consistently shows times that are 30-60 ahead of the MTA countdown clocks?

**Comment by jamespfennell** [comment]
The bug is fixed now, thanks again for the report!

I think being ahead of the MTA clocks may be by design. The app uses each train's arrival time rather than its departure time (these are typically 15 seconds apart) and always rounds _down_ to the nearest minute. So if the train is arriving in 1 min 20 seconds and departing in 1 min 35 seconds, the app will take 1 min 20 seconds and round it down to 1 minute. I suspect the countdown clocks might be taking the 1 min 35 seconds departure time and rounding it to the nearest minute, and so they report 2 minutes.

My motivation was that I use the app to ensure I get to the station on time, and so I never want the app to give me a time that is too late. I.e., if the train is 1 min 35 seconds away and I round to 2 minutes, even if I arrive in 2 minutes I will miss the train. This is why the app rounds down and uses the earlier arrival time. But maybe this is too conservative and makes the app slightly inaccurate?

**Comment by lia-simeone** [comment]
Fascinating. Totally get the rationale from a design standpoint. Let me keep an eye on it for a bit and see how I feel. 

I think that logic largely makes sense but I have noticed once or twice that I _don't_ hustle to get to the train because I see the next one is in 1 minute and I can't make it. But if it's actually 1 minute 35 seconds away, I stroll into the station, realize I could have made it if I had walked faster and then hate myself haha. But this probably doesn't happen super often.
